### PR TITLE
test: integration AuthFetch and auth middleware; fix: middleware not passing integration tests.

### PR DESCRIPTION
**Pull request by dorzepowski**

### CODE REVIEW

File: pkg/middleware/auth/middleware.go [hunk 1]
```diff
@@ -37,29 +35,17 @@ func New(cfg Config) (*Middleware, error) {
 		return nil, errors.New("wallet is required")
 	}
```
[the remark this answers is not in the document]
Comment (dorzepowski) [reply pkg/middleware/auth/middleware.go:37]: No, it's not - logging.Child have nil check for logger

File: pkg/transport/http/transport.go [hunk 2]
```diff
@@ -54,30 +53,31 @@ func (r *ResponseRecorder) GetBody() []byte {
...
-	r.body = append(r.body, b...)
 	return len(b), nil
 }
 
-// Flush writes the response header and body if they have not been written yet.
+// Flush writes the response header and body from the ResponseRecorder to the underlying http.ResponseWriter.
 func (r *ResponseRecorder) Flush() error {
-	if r.statusCode == 0 {
-		r.statusCode = http.StatusOK
+	if r.statusCode < 100 {
```
Comment (chris-4chain) [inline pkg/transport/http/transport.go]: Shouldn't be checked also the upper limit?
```go
statusCode < 100 || statusCode > 999
```

File: pkg/transport/http/transport.go [hunk 3]
```diff
@@ -187,20 +187,20 @@ func (t *Transport) Send(ctx context.Context, message *auth.AuthMessage) error {
 
 	switch message.MessageType {
 	case auth.MessageTypeInitialResponse, auth.MessageTypeCertificateResponse:
-		resp.Header().Set(constants.HeaderVersion, message.Version)
-		resp.Header().Set(constants.HeaderMessageType, string(message.MessageType))
-		resp.Header().Set(constants.HeaderIdentityKey, message.IdentityKey.ToDERHex())
+		resp.Header().Set(brc104.HeaderVersion, message.Version)
```
Comment (chris-4chain) [inline pkg/transport/http/transport.go:190]: **Just saying**
For this switch case, helper functions for each case would increase readability.

File: pkg/transport/http/transport.go [hunk 4]
```diff
@@ -220,33 +220,33 @@ func (t *Transport) Send(ctx context.Context, message *auth.AuthMessage) error {
...
-			resp.Header().Set(constants.HeaderSignature, hex.EncodeToString(message.Signature))
+			resp.Header().Set(brc104.HeaderSignature, hex.EncodeToString(message.Signature))
 		}
 
-		resp.Header().Set(constants.HeaderRequestID, requestID)
+		resp.Header().Set(brc104.HeaderRequestID, requestID)
 
 		if t.wallet != nil {
-			peerIdentityKeyStr := req.Header.Get(constants.HeaderIdentityKey)
+			peerIdentityKeyStr := req.Header.Get(brc104.HeaderIdentityKey)
```
Comment (chris-4chain) [inline pkg/transport/http/transport.go:249]: Isn't the case where `req` is nil? Look a few lines above - there is a check `req != nil` - so, I guess the `req` variable can be nil.

File: pkg/transport/http/transport.go [hunk 5]
```diff
@@ -317,56 +324,64 @@ func ParseAuthMessageFromRequest(req *http.Request) (*AuthMessageWithRequestID,
...
-
-		message.Signature = sigBytes
 	}
+
 	msg := &AuthMessageWithRequestID{
-		AuthMessage: message,
-		RequestID:   requestID,
+		RequestID:      requestID,
+		RequestIDBytes: requestIDBytes,
+		AuthMessage: &auth.AuthMessage{
```
Comment (chris-4chain) [inline pkg/transport/http/transport.go:375]: You haven't initialized two fields of the `AuthMessage`: `InitialNonce` and `Certificates` - is it done on purpose?

File: pkg/internal/integrationtests/auth_middleware_integration_test.go [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+			cleanup := given.Server().WithMiddleware(authMiddleware).
+				WithRoute("/", func(w http.ResponseWriter, r *http.Request) {
+					then.Request(r).
+						HasMethod(test.method).
+						HasPath(test.path).
+						HasQueryMatching(url.PathEscape(test.query)).
+						HasHeadersContaining(test.headers).
+						HasBody(test.body)
+
+					// TODO check identity key
```
Comment (chris-4chain) [inline pkg/internal/integrationtests/auth_middleware_integration_test.go:167]: I guess, you are aware of this TODO and you'll address it later?

File: pkg/internal/integrationtests/auth_middleware_integration_test.go [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+
+func TestAuthMiddlewareAndAuthFetchIntegration(t *testing.T) {
+	testCases := map[string]struct {
+		path    string
+		method  string
+		headers map[string]string
+		query   string
+		body    string
+	}{
+		"get request on server root": {
```
Comment (chris-4chain) [inline pkg/internal/integrationtests/auth_middleware_integration_test.go:23]: AI suggested me that following cases are not covered. 
Check it out, maybe subset of them are interesting:

HTTP method coverage
- HEAD requests (root, path, query, headers, path+query+headers).
- PUT, PATCH, DELETE variants similar to your POST matrix (with/without body, with Content-Type, with headers, with query, path+query+headers+body).
- OPTIONS preflight (with Access-Control-Request-Method and Access-Control-Request-Headers).

Combinations you have for POST but not covered in some permutations
- POST with query params only (no body, no headers).
- POST on path with query params only.
- GET with query params and headers (root).
- GET on path with query params and headers.

Headers
- Multiple values for the same header (e.g., repeated Set-Cookie-like scenarios or custom headers with multiple values).
- Header name case-insensitivity (e.g., "authorization" vs "Authorization", "x-bsv-test" vs "X-Bsv-Test").
- Mixed custom X-BSV headers (more than one custom X-Bsv-* header).
- Accept header variations to validate error content-negotiation (text/plain vs default JSON) on failure paths.

Content types and bodies
- Content-Type with parameters (application/json; charset=utf-8).
- Non-JSON bodies with mismatched Content-Type (e.g., Content-Type: application/json but body is plain text).
- application/octet-stream and multipart/form-data.
- Empty vs nil body distinction.

Query strings and paths
- Encoded characters and edge cases (spaces, plus, percent-encoding).
- Duplicate keys (test=1&test=2), empty value (test=), key without value.
- Trailing slashes and deeper subpaths (/ping/, /ping/sub).

Authentication paths and errors
- Unauthenticated requests when middleware is configured to disallow unauthenticated (expect 401).
- Error mapping cases: invalid/expired session (401), invalid nonce (400), invalid message (400), missing certificate types (400 with aggregated message), invalid signature path.
- Ensure x-bsv-auth-identity-key presence/absence under unauthenticated-allowed vs disallowed modes.

Other robustness
- Very large body payloads.
- Cookies present.
- Concurrency: multiple simultaneous requests with same/different clients to the same user (you cover subsequent, not concurrent).
- HEAD-specific behavior (no body in response, but headers present).

File: pkg/internal/integrationtests/auth_middleware_integration_test.go [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+			// when:
+			serverURL := given.Server().URL()
+			serverURL.Path = test.path
+			serverURL.RawQuery = test.query
+
+			response, err := httpClient.Fetch(t.Context(), serverURL.String(), &clients.SimplifiedFetchRequestOptions{
+				Method:       test.method,
+				Headers:      test.headers,
+				Body:         []byte(test.body),
+				RetryCounter: to.Ptr(1),
```
Comment (chris-4chain) [inline pkg/internal/integrationtests/auth_middleware_integration_test.go:191]: Do you have a test for `RetryCounter` higher-than-one ?

File: pkg/internal/testabilities/fixture_client.go [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+	logger *slog.Logger
+}
+
+func WithClientLogger(logger *slog.Logger) func(options *ClientFixtureOptions) {
+	return func(options *ClientFixtureOptions) {
+		options.logger = logger
+	}
+}
+
+func WithoutLoggingFromClient() func(*ClientFixtureOptions) {
```
Comment (chris-4chain) [inline pkg/internal/testabilities/fixture_client.go]: It seems unused

File: pkg/internal/testabilities/fixture_middleware.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+	logger *slog.Logger
+}
+
+func WithMiddlewareLogger(logger *slog.Logger) func(options *MiddlewareFixtureOptions) {
+	return func(options *MiddlewareFixtureOptions) {
+		options.logger = logger
+	}
+}
+
+func WithoutLoggingFromMiddleware() func(*MiddlewareFixtureOptions) {
```
Comment (chris-4chain) [inline pkg/internal/testabilities/fixture_middleware.go]: Unused - but perhaps you want to keep it for future

File: pkg/internal/testabilities/testabilities_options.go [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+	logger *slog.Logger
+}
+
+func WithLogger(logger *slog.Logger) func(*Options) {
+	return func(options *Options) {
+		options.logger = logger
+	}
+}
+
+func WithoutLogging() func(*Options) {
```
Comment (chris-4chain) [inline pkg/internal/testabilities/testabilities_options.go]: Both functions in this file are unused.

File: pkg/transport/http/transport.go (hunk 2 above)
Comment (dorzepowski) [reply pkg/transport/http/transport.go]: actually, I will check only for status == 0,
because other cases are checked when doing WriteHeader, and WriteHeader is the only way to set the status code

File: pkg/transport/http/transport.go (hunk 3 above)
Comment (dorzepowski) [reply pkg/transport/http/transport.go:190]: I will have a big refactoring PR, this code will also change there, so let's leave it for later.

File: pkg/transport/http/transport.go (hunk 4 above)
Comment (dorzepowski) [reply pkg/transport/http/transport.go:249]: I will have a big refactoring PR - where I added the function to get the request, which is also returning error if it is missing.

File: pkg/transport/http/transport.go (hunk 5 above)
Comment (dorzepowski) [reply pkg/transport/http/transport.go:375]: 1.  InitialNonce - as both tests (integration and regression) are passing I assume it's correct - and from what I remember, this field is filled by Peer struct by retrieving the value from session (and there is no header in constants that could hold this value, so this confirms my assumptions additionally)
2. Certificates - I'm not sure - I'm not handling (testing) certificates yet - but because there is no constant for header name with certificates I assume that certificates cannot be passed via header, so there is no option to pass them with general message.

File: pkg/internal/integrationtests/auth_middleware_integration_test.go (hunk 6 above)
Comment (dorzepowski) [reply pkg/internal/integrationtests/auth_middleware_integration_test.go:167]: yes, in the 4th PR:
Just to explain the plan of PRs:
1. integration tests
2. regression tests with grpc server - to have option to test also several calls of the same typescript authfetch
3. refactoring
4. handle identity key

File: pkg/internal/integrationtests/auth_middleware_integration_test.go (hunk 7 above)
Comment (dorzepowski) [reply pkg/internal/integrationtests/auth_middleware_integration_test.go:23]: Added task for improving tests coverage later:
https://github.com/bsv-blockchain/go-bsv-middleware/issues/140

File: pkg/internal/integrationtests/auth_middleware_integration_test.go (hunk 8 above)
Comment (dorzepowski) [reply pkg/internal/integrationtests/auth_middleware_integration_test.go:191]: No, but it is AuthFetch feature, not the middleware feature 
And AuthFetch is not a part of this repository.

**Review by chris-4chain — APPROVED** [review]

**Review by jakubmkowalski — APPROVED** [review]
